dsp_permissions_scripts/oap/oap_get.py

```diff
@@ -30,6 +33,82 @@
     "knora-api:userHasPermission",
     "knora-api:hasPermissions",
 ]
+KB_RESCLASSES = [
+    "knora-api:VideoSegment",
+    "knora-api:AudioSegment",
+    "knora-api:Region",
+    "knora-api:Annotation",
+    "knora-api:LinkObj",
+]
+
+
+def get_oaps_of_kb_resclasses(dsp_client: DspClient, project_iri: str, oap_config: OapRetrieveConfig) -> list[Oap]:
+    match oap_config.retrieve_resources:
+        case "specified_res_classes":
+            specified_kb_resclasses = [x for x in KB_RESCLASSES if x in oap_config.specified_res_classes]
+            res_only_oaps = _get_oaps_of_specified_kb_resclasses(dsp_client, project_iri, specified_kb_resclasses)
+        case "all":
+            res_only_oaps = _get_oaps_of_specified_kb_resclasses(dsp_client, project_iri, KB_RESCLASSES)
+
+    match oap_config.retrieve_values:
+        case "none":
+            return res_only_oaps
+        case "specified_props":
+            enriched_oaps = _enrich_with_value_oaps(dsp_client, res_only_oaps, oap_config.specified_props)
+        case "all":
+            enriched_oaps = _enrich_with_value_oaps(dsp_client, res_only_oaps)
+
+    return enriched_oaps
+
+
+def _get_oaps_of_specified_kb_resclasses(
+    dsp_client: DspClient, project_iri: str, kb_resclasses: list[str]
+) -> list[Oap]:
+    all_oaps: list[Oap] = []
+    for resclass in kb_resclasses:
+        all_oaps.extend(_get_oaps_of_one_kb_resclass(dsp_client, project_iri, resclass))
+    return all_oaps
+
+
+def _enrich_with_value_oaps(
+    dsp_client: DspClient, res_only_oaps: list[Oap], restrict_to_props: list[str] | None = None
+) -> list[Oap]:
+    complete_oaps = copy.deepcopy(res_only_oaps)
+    for oap in complete_oaps:
+        full_resource = dsp_client.get(f"/v2/resources/{quote_plus(oap.resource_oap.resource_iri)}")
+        oap.value_oaps = _get_value_oaps(full_resource, restrict_to_props)
+    return complete_oaps
+
+
+def _get_oaps_of_one_kb_resclass(dsp_client: DspClient, project_iri: str, resclass: str) -> list[Oap]:
+    oaps: list[Oap] = []
+    mayHaveMoreResults: bool = True
+    offset = 0
+    while mayHaveMoreResults:
+        sparql_query = """
+        PREFIX knora-api: <http://api.knora.org/ontology/knora-api/v2#>
+
+        CONSTRUCT {
+            ?kb_resclass knora-api:isMainResource true .
+        } WHERE {
+            BIND(<%(project_iri)s> as ?project_iri) .
+            ?kb_resclass a %(resclass)s .
+            ?kb_resclass knora-api:attachedToProject ?project_iri .
+        }
+        OFFSET %(offset)s
+        """ % {"resclass": resclass, "project_iri": project_iri, "offset": offset}  # noqa: UP031 (printf-string-formatting)
+        sparql_query_stripped = re.sub(r"\s+", " ", sparql_query).strip()
+        if not (response := dsp_client.get(f"/v2/searchextended/{quote(sparql_query_stripped, safe='')}")):
```

Comment:
That's what I wanted to do first, but then I stumbled over 2 difficulties:

1.  I didn't find out which `Content-Type` to use. After several attempts, it didn't work out, then I gave up.
2. The current architecture of the `DspClient` class is built upon the assumption that the data that is sent in a POST request is a python dictionary. Changing this would produce a big diff.

Reply:
1. In all my local gravsearch post requests, I have never specified the content type, and it always worked. Maybe not specifying it would do the trick?
2. Maybe this is an indication that a generic solution isn't always ideal - if you had a `GravsearchClient`, this wouldn't be an issue.

But if it's easier to keep it like this, then I'm not strongly opposed.

Reply:
Though, writing this now, this makes me think that a `GravsearchClient` would actually be a neat way of hiding away the accidental complexity of dealing with paging. If you think about it, paging should not be a concern of a `get_oaps_...` method, it is lover level and an implementation detail of gravsearch.

Reply:
Sure, go ahead like this.

But just as a side note: For DSP-TOOLS we have already agreed that we will split up the connection into less generic and more specific ones. And we agreed on that because we came to the conclusion that the mechanisms of retry and logging can still remain generic, and be re-used by various connections/clients. It's not about promoting code duplication, it's about finding the right granularity for reuseable components.

Reply:
Regarding DSP-TOOLS: Sure, I forgot that.